src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -610,27 +618,40 @@ def _prepare_for_mds_upgrade(
 
             # scale down this filesystem?
             if mdsmap["max_mds"] > 1:
-                self.mgr.log.info('Upgrade: Scaling down filesystem %s' % (
-                    fs_name
-                ))
-                if fscid not in self.upgrade_state.fs_original_max_mds:
-                    self.upgrade_state.fs_original_max_mds[fscid] = mdsmap['max_mds']
-                    self._save_upgrade_state()
-                ret, out, err = self.mgr.check_mon_command({
-                    'prefix': 'fs set',
-                    'fs_name': fs_name,
-                    'var': 'max_mds',
-                    'val': '1',
-                })
-                continue_upgrade = False
-                continue
+                if self.upgrade_state.do_not_reduce_mds:
+                    self.mgr.log.info('Upgrade: Not scaling down filesystem '
+                                      'as do_not_reduce_mds is set to true')
+                    ret, out, err = self.mgr.check_mon_command({
+                        'prefix': 'fs fail',
+                        'fs_name': fs_name
+                    })
+                    continue_upgrade = True
```

Comment:
You need to check if the file system is already "not joinable" **and** no mds in `mdsmap['up']` otherwise, in a multifs cluster, this could run `fs fail` multiple times needlessly.

Reply:
I see fs data is being extracted using `self.mgr.get("fs_map")` everywhere in the code, assuming it as a dictionary, what value do I need to use to check if the fs is already "not joinable"? If I want to check for the all the values of the key `fs_map`, where can i find it?

Reply:
I've updated the code.